File: 2023/src/bin/day_17.py
# pylint: disable=missing-module-docstring,missing-class-docstring,missing-function-docstring

from dataclasses import dataclass
from enum import IntEnum
from queue import PriorityQueue


def read_input() -> list[list[int]]:
    with open("input/day17.txt", encoding="utf-8") as f:
        return [[int(c) for c in line.strip()] for line in f.readlines()]


class Direction(IntEnum):
    UP = 0
    RIGHT = 1
    DOWN = 2
    LEFT = 3

    def turn_left(self) -> "Direction":
        return Direction((self.value - 1) % 4)

    def turn_right(self) -> "Direction":
        return Direction((self.value + 1) % 4)

    def mv(
        self, rows: int, cols: int, row: int, col: int, length: int
    ) -> tuple[int, int] | None:
        match self:
            case Direction.UP:
                if row >= length:
                    return row - length, col
            case Direction.RIGHT:
                if col + length < cols:
                    return row, col + length
            case Direction.DOWN:
                if row + length < rows:
                    return row + length, col
            case Direction.LEFT:
                if col >= length:
                    return row, col - length
        return None


@dataclass(order=True)
class State:
    row: int
    col: int
    direction: Direction
    path_cost: int


def mv(
    in_map: list[list[int]], min_moves: int, max_moves: int, state: State
) -> list[State]:
    next_states: list[State] = []
    rows = len(in_map)
    cols = len(in_map[0])
    left = state.direction.turn_left()
    right = state.direction.turn_right()
    left_cost = state.path_cost
    right_cost = state.path_cost
    for length in range(1, max_moves + 1):
        mv_left = left.mv(rows, cols, state.row, state.col, length)
        mv_right = right.mv(rows, cols, state.row, state.col, length)
        if mv_left is not None:
            next_row, next_col = mv_left
            left_cost += in_map[next_row][next_col]
            if length >= min_moves:
                next_states.append(State(next_row, next_col, left, left_cost))
        if mv_right is not None:
            next_row, next_col = mv_right
            right_cost += in_map[next_row][next_col]
            if length >= min_moves:
                next_states.append(State(next_row, next_col, right, right_cost))

    return next_states


def calculate_priority(heuristic_map: list[list[int]], state: State) -> int:
    return state.path_cost + heuristic_map[state.row][state.col]


def hash_position(in_map: list[list[int]], state: State) -> int:
    return (state.row * len(in_map[0]) + state.col) * 4 + state.direction.value


def precalculate_heuristic(in_map: list[list[int]]) -> list[list[int]]:
    rows = len(in_map)
    cols = len(in_map[0])
    to_process: PriorityQueue[(int, int, int)] = PriorityQueue()
    to_process.put((0, rows - 1, cols - 1))
    results: list[list[int | None]] = [[None for _ in range(cols)] for _ in range(rows)]
    to_find = len(results * len(results[0]))
    while to_find > 0:
        (cost, row, col) = to_process.get()
        if results[row][col] is not None:
            continue
        results[row][col] = cost
        to_find -= 1
        next_cost = cost + in_map[row][col]
        next_states = [
            (next_cost, row + r, col + c)
            for (r, c) in [(0, 1), (1, 0), (0, -1), (-1, 0)]
            if row + r >= 0 and row + r < rows and col + c >= 0 and col + c < cols
        ]
        for next_state in next_states:
            if results[next_state[1]][next_state[2]] is None:
                to_process.put(next_state)
    assert all(result is not None for row in results for result in row)
    return results


def run(in_map: list[list[int]], min_moves: int, max_moves: int) -> int:
    heuristic_map = precalculate_heuristic(in_map)
    initial_states = [
        State(0, 0, direction, 0) for direction in [Direction.RIGHT, Direction.DOWN]
    ]
    been = [False for _ in range(len(in_map) * len(in_map[0]) * 4)]
    to_process: PriorityQueue[(int, State)] = PriorityQueue()
    for s in initial_states:
        to_process.put((calculate_priority(heuristic_map, s), s))

    while True:
        (_, state) = to_process.get()
        if state.row == len(in_map) - 1 and state.col == len(in_map[0]) - 1:
            return state.path_cost
        if been[hash_position(in_map, state)]:
            continue
        been[hash_position(in_map, state)] = True
        next_states = mv(in_map, min_moves, max_moves, state)
        for next_state in next_states:
            if not been[hash_position(in_map, next_state)]:
                to_process.put(
                    (calculate_priority(heuristic_map, next_state), next_state)
                )


def main():
    in_map = read_input()
    part_1 = run(in_map, 1, 3)
    print("Part 1:", part_1)
    part_2 = part_2 = run(in_map, 4, 10)
    print("Part 2:", part_2)


if __name__ == "__main__":
    main()

File: 2023/src/bin/day_17.rs
use priority_queue::PriorityQueue;
use util::PerfTimer;

fn input() -> Vec<Vec<u32>> {
    util::get_day_input(17)
        .lines()
        .map(|line| line.chars().map(|c| c.to_digit(10).unwrap()).collect())
        .collect()
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
enum Direction {
    Up,
    Right,
    Down,
    Left,
}

impl Direction {
    fn turn_left(self) -> Self {
        match self {
            Direction::Up => Direction::Left,
            Direction::Right => Direction::Up,
            Direction::Down => Direction::Right,
            Direction::Left => Direction::Down,
        }
    }

    fn turn_right(self) -> Self {
        match self {
            Direction::Up => Direction::Right,
            Direction::Right => Direction::Down,
            Direction::Down => Direction::Left,
            Direction::Left => Direction::Up,
        }
    }

    fn mv(
        self,
        rows: usize,
        cols: usize,
        row: usize,
        col: usize,
        len: usize,
    ) -> Option<(usize, usize)> {
        match self {
            Direction::Up => {
                if row >= len {
                    Some((row - len, col))
                } else {
                    None
                }
            }
            Direction::Right => {
                if col + len < cols {
                    Some((row, col + len))
                } else {
                    None
                }
            }
            Direction::Down => {
                if row + len < rows {
                    Some((row + len, col))
                } else {
                    None
                }
            }
            Direction::Left => {
                if col >= len {
                    Some((row, col - len))
                } else {
                    None
                }
            }
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
struct State {
    row: usize,
    col: usize,
    direction: Direction,
    path_cost: u32,
}

fn mv(map: &[Vec<u32>], min_moves: usize, max_moves: usize, state: State) -> Vec<State> {
    let mut next_states = Vec::new();
    let rows = map.len();
    let cols = map[0].len();
    let left = state.direction.turn_left();
    let right = state.direction.turn_right();
    let mut left_cost = state.path_cost;
    let mut right_cost = state.path_cost;
    for len in 1..=max_moves {
        if let Some((next_row, next_col)) = left.mv(rows, cols, state.row, state.col, len) {
            left_cost += map[next_row][next_col];
            if len >= min_moves {
                next_states.push(State {
                    row: next_row,
                    col: next_col,
                    path_cost: left_cost,
                    direction: left,
                })
            }
        }
        if let Some((next_row, next_col)) = right.mv(rows, cols, state.row, state.col, len) {
            right_cost += map[next_row][next_col];
            if len >= min_moves {
                next_states.push(State {
                    row: next_row,
                    col: next_col,
                    path_cost: right_cost,
                    direction: right,
                })
            }
        }
    }
    next_states
}

fn calculate_priority(map: &[Vec<u32>], state: &State) -> u32 {
    u32::MAX
        - state.path_cost
        - ((map.len() - state.row - 1) + (map[0].len() - state.col - 1)) as u32
}

fn hash_position(map: &[Vec<u32>], state: &State) -> usize {
    (state.row * map[0].len() + state.col) * 4 + state.direction as usize
}

fn run(map: &[Vec<u32>], min_moves: usize, max_moves: usize) -> u32 {
    let initial_states: Vec<State> = [Direction::Right, Direction::Down]
        .into_iter()
        .map(|direction| State {
            row: 0,
            col: 0,
            direction,
            path_cost: 0,
        })
        .collect();
    let mut been = vec![false; map.len() * map[0].len() * 4];
    let mut to_process = PriorityQueue::new();
    for s in initial_states {
        to_process.push(s, calculate_priority(map, &s));
    }
    loop {
        let (state, _) = to_process.pop().unwrap();
        if state.row == map.len() - 1 && state.col == map[0].len() - 1 {
            break state.path_cost;
        }
        if been[hash_position(map, &state)] {
            continue;
        };
        been[hash_position(map, &state)] = true;
        let next_states = mv(map, min_moves, max_moves, state);
        for next_state in next_states {
            if !been[hash_position(map, &next_state)] {
                to_process.push(next_state, calculate_priority(map, &next_state));
            }
        }
    }
}

fn main() {
    let map = input();
    {
        let _timer = PerfTimer::new("Part 1");
        let part_1 = run(&map, 1, 3);
        println!("Part 1: {part_1}");
    }
    {
        let _timer = PerfTimer::new("Part 2");
        let part_2 = run(&map, 4, 10);
        println!("Part 2: {part_2}");
    }
}
